integration-tests/src/upgrade/tester.rs

```diff
@@ -0,0 +1,352 @@
+            !tester
+                .l1_provider
+                .get_code_at(bytecode_supplier_address)
+                .await?
+                .is_empty(),
+            "Bytecode supplier contract is not deployed at expected address {bytecode_supplier_address:?}; if zkos-l1-state.json was updated, update the address in the test code"
+        );
+        let bytecode_supplier = interfaces::BytecodesSupplier::new(
+            bytecode_supplier_address,
+            tester.l1_provider.clone(),
+        );
+
+        Ok(Self {
+            tester,
+            bridgehub,
+            bridgehub_owner,
+            ctm,
+            ctm_owner,
+            diamond_proxy,
+            diamond_proxy_admin,
+            l1_chain_admin,
+            l1_chain_admin_owner,
+            bytecode_supplier,
+            protocol_version,
+        })
+    }
+
+    async fn enable_impersonation(&self) -> anyhow::Result<()> {
+        // Enable impersonation and fund all governance accounts
+        for addr in [
+            self.bridgehub_owner,
+            self.ctm_owner,
+            self.diamond_proxy_admin,
+            self.l1_chain_admin_owner,
+        ] {
+            self.tester
+                .l1_provider
+                .anvil_impersonate_account(addr)
+                .await?;
+            self.tester
+                .l1_provider
+                .send_transaction(
+                    TransactionRequest::default()
+                        .with_to(addr)
+                        .with_value(U256::from(10).pow(U256::from(18u64))), // 1 ETH
+                )
+                .await?
+                .expect_successful_receipt()
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn fund_wallets(&self) -> anyhow::Result<()> {
+        for wallet in [self.bridgehub_owner] {
+            self.tester
+                .l1_provider
+                .send_transaction(
+                    TransactionRequest::default()
+                        .with_to(wallet)
+                        .with_value(U256::from(10).pow(U256::from(18u64))), // 1 ETH
+                )
+                .await?
+                .expect_successful_receipt()
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn wait_for_genesis_upgrade(&self) -> anyhow::Result<()> {
+        // Wait until the genesis upgrade is executed
+        while self.diamond_proxy.getTotalBatchesExecuted().call().await? == U256::ZERO {
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        }
+        Ok(())
+    }
+
+    pub async fn pause_bridgehub_migrations(&self) -> anyhow::Result<()> {
+        let pause_migration_tx = self
+            .bridgehub
+            .pauseMigration()
+            .into_transaction_request()
+            .with_from(self.bridgehub_owner);
+        let hash = self
+            .tester
+            .l1_provider
+            .anvil_send_impersonated_transaction(pause_migration_tx)
```

Comment:
What does this do exactly? Shouldn't regular send be okay here since you impersonated the account already?

Reply:
Encapsulated this logic and added a comment to explain purpose.